### PR TITLE
feat!(db): incorporate detection tag

### PR DESCRIPTION
**Pull request by shino**
cf https://github.com/MaineK00n/vuls-data-update/pull/267

### CODE REVIEW

File: pkg/db/common/boltdb/boltdb.go [hunk 1]
```diff
@@ -698,7 +698,10 @@ func putDetection(tx *bolt.Tx, data dataTypes.Data) error {
 			return errors.Wrapf(err, "create bucket: %s if not exists", fmt.Sprintf("%s -> index", d.Ecosystem))
 		}
 
-		pkgs := util.WalkCriteria(d.Criteria)
+		pkgs := make([]string, 0)
```
Comment (MaineK00n) [inline pkg/db/common/boltdb/boltdb.go:701]:
```suggestion
		var pkgs []string
```

File: pkg/db/common/redis/redis.go [hunk 2]
```diff
@@ -675,7 +675,10 @@ func (c *Connection) putDetection(ctx context.Context, data dataTypes.Data) erro
 			return errors.Wrapf(err, "HSET %s %s %q", fmt.Sprintf("%s#detection#%s", d.Ecosystem, data.ID), data.DataSource, string(bs))
 		}
 
-		pkgs := util.WalkCriteria(d.Criteria)
+		pkgs := make([]string, 0)
```
Comment (MaineK00n) [inline pkg/db/common/redis/redis.go:678]:
```suggestion
		var pkgs []string
```

File: pkg/detect/ospkg/ospkg.go [hunk 3]
```diff
@@ -129,28 +132,29 @@ func Detect(dbc db.DB, sr scanTypes.ScanResult) (detectTypes.VulnerabilityDataDe
 		}
 	}
 
-	contents := make(map[string]map[sourceTypes.SourceID]criteriaTypes.FilteredCriteria)
+	contents := make(map[string]map[sourceTypes.SourceID]conditionTypes.FilteredCondition)
 	for rootID, m := range cm {
-		for sourceID, fca := range m {
-			qs := make([]criterionTypes.Query, 0, len(fca.indexes))
-			for _, idx := range fca.indexes {
+		for sourceID, fs := range m {
```
Comment (MaineK00n) [inline pkg/detect/ospkg/ospkg.go:137]: It is better to use either `fs` or `containedList`
https://github.com/MaineK00n/vuls2/pull/84/files#diff-bf1f3d3fc4dd9c7172fda625d2f706506db8624e551ebb6e0adab81affe5dd34R83

File: pkg/detect/ospkg/ospkg.go [hunk 4]
```diff
@@ -172,20 +176,20 @@ func replaceIndexes(ac criteriaTypes.FilteredCriteria, indexes []int) criteriaTy
...
+		for _, a := range cn.Accepts {
 			is = append(is, indexes[a])
 		}
-		co.Accepts = is
-		cos = append(cos, co)
+		cn.Accepts = is
+		cns = append(cns, cn)
 	}
-	replaced.Criterions = cos
+	replaced.Criterions = cns
```
Comment (MaineK00n) [inline pkg/detect/ospkg/ospkg.go:192]: You should make the same change here too.
https://github.com/MaineK00n/vuls2/pull/84/files#diff-bf1f3d3fc4dd9c7172fda625d2f706506db8624e551ebb6e0adab81affe5dd34R125-R138

File: pkg/detect/types/detect.go [hunk 5]
```diff
@@ -30,6 +30,21 @@ type VulnerabilityData struct {
 }
 
 type VulnerabilityDataDetection struct {
-	Ecosystem ecosystemTypes.Ecosystem                                           `json:"ecosystem,omitempty"`
-	Contents  map[string]map[sourceTypes.SourceID]criteriaTypes.FilteredCriteria `json:"contents,omitempty"`
+	Ecosystem ecosystemTypes.Ecosystem                                             `json:"ecosystem,omitempty"`
+	Contents  map[string]map[sourceTypes.SourceID]conditionTypes.FilteredCondition `json:"contents,omitempty"`
+}
+
+func (r *DetectResult) Sort() {
```
Comment (MaineK00n) [inline pkg/detect/types/detect.go:37]: In what situations is this sort intended to be used?

File: pkg/detect/ospkg/ospkg.go (hunk 4 above)
Comment (shino) [reply pkg/detect/ospkg/ospkg.go:192]: done https://github.com/MaineK00n/vuls2/compare/60ef7c3d8f84777e42828085dd0746e008187ee2..296a7fd45df1a2a7c22523a475d4b19cb089b837

File: pkg/detect/ospkg/ospkg.go (hunk 3 above)
Comment (shino) [reply pkg/detect/ospkg/ospkg.go]: Nice catch!
https://github.com/MaineK00n/vuls2/compare/be79505f64c1bf5aa71a1182d2a8e0509eceea5a..4493ac52c8e78553815338260e823fa5b18ef277

File: pkg/detect/types/detect.go (hunk 5 above)
Comment (shino) [reply pkg/detect/types/detect.go:37]: Not needed yet. will remove

Comment (shino) [reply pkg/detect/types/detect.go]: https://github.com/MaineK00n/vuls2/compare/296a7fd45df1a2a7c22523a475d4b19cb089b837..edd42367d12bf3ac9083eb390e4cfb4e0271ace6

File: pkg/db/common/redis/redis.go [hunk 6]
```diff
@@ -665,8 +664,8 @@ func (c *Connection) PutVulnerabilityData(root string) error {
 }
 
 func (c *Connection) putDetection(ctx context.Context, data dataTypes.Data) error {
-	for _, d := range data.Detection {
-		bs, err := util.Marshal(d.Criteria)
+	for _, d := range data.Detections {
+		bs, err := util.Marshal(d.Conditions)
 		if err != nil {
 			return errors.Wrap(err, "marshal criteria")
```
Comment (MaineK00n) [inline pkg/db/common/redis/redis.go:670]:
```suggestion
			return errors.Wrap(err, "marshal conditions")
```

File: pkg/db/common/redis/redis.go [hunk 7]
```diff
@@ -506,24 +505,24 @@ func (c *Connection) getDetection(ctx context.Context, rootID string) ([]dbTypes
...
 			if err != nil {
 				return nil, errors.Wrapf(err, "as bytes %s", fmt.Sprintf("%s -> %s", fmt.Sprintf("%s#detection#%s", ecosystem, rootID), k))
 			}
 
-			var c criteriaTypes.Criteria
-			if err := util.Unmarshal(bs, &c); err != nil {
+			var conds []conditionTypes.Condition
+			if err := util.Unmarshal(bs, &conds); err != nil {
 				return nil, errors.Wrapf(err, "unmarshal %s", fmt.Sprintf("%s -> %s", fmt.Sprintf("%s#detection#%s", ecosystem, rootID), k))
 			}
```
Comment (MaineK00n) [inline pkg/db/common/redis/redis.go:519]:
```suggestion
```

**Review by MaineK00n — DISMISSED** [review]

**Review by MaineK00n — APPROVED** [review]

**Pull request merged**